commit e49157f3cfc3d6f4da40be13fc24557c23b715bb
Author:     Yan Cheng <58191769+yanchengnv@users.noreply.github.com>
AuthorDate: Thu Jul 24 12:15:10 2025 -0400
Commit:     GitHub <noreply@github.com>
CommitDate: Thu Jul 24 12:15:10 2025 -0400

    Fix max client limit to exclude admin clients (#3552)
    
    Fixes # .
    
    ### Description
    
    This PR fixes the bug 4 (reported in
    https://nvbugspro.nvidia.com/bug/5390831) that failed to allow admin
    clients once max number of FL clients are connected. Admin clients will
    not be counted as FL clients.
    
    This PR also attempts to fix bug 5 in the report. Though we couldn't
    reproduce the issue following QA instructions, we found and fixed a
    place that could potentially cause this.
    
    This PR also fixes bug 6.
    
    
    ### Types of changes
    <!--- Put an `x` in all the boxes that apply, and remove the not
    applicable items -->
    - [x] Non-breaking change (fix or new feature that would not break
    existing functionality).
    - [ ] Breaking change (fix or new feature that would cause existing
    functionality to change).
    - [ ] New tests added to cover the changes.
    - [ ] Quick tests passed locally by running `./runtest.sh`.
    - [ ] In-line docstrings updated.
    - [ ] Documentation updated.
---
 nvflare/private/fed/server/server_engine.py | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index b2d659da..80122282 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -126,6 +126,12 @@ class ServerEngine(ServerEngineInternalSpec, StreamableEngine):
 
         self.kv_list = parse_vars(args.set)
 
+    def has_relays(self):
+        if not self.client_manager:
+            return False
+        else:
+            return self.client_manager.has_relays()
+
     def _get_run_folder(self, job_id):
         workspace = Workspace(self.args.workspace)
         return workspace.get_run_dir(job_id)
